File: apps/api/src/utils/trace-context.ts
/**
 * traceparent 上下文工具(2026-07-22 立,跨服务 traceparent 透传)。
 *
 * W3C Trace Context 格式:version-trace_id-parent_id-flags
 * 例:00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01
 *
 * 核心能力:
 * - generateTraceparent():生成新 traceparent(新 trace)
 * - getTraceparentFromRequest():从 Fastify request 头解析
 * - propagateToHeaders():把 traceparent 注入到出站请求 headers(api → ai-service)
 * - extractTraceId():解析 trace_id(32 hex)
 */

import { randomBytes } from 'node:crypto'
import type { FastifyRequest } from 'fastify'

export interface TraceContext {
  traceId: string      // 32 hex
  parentId: string     // 16 hex
  version: string      // 2 hex(通常 "00")
  flags: string        // 2 hex(通常 "01" sampled)
}

/**
 * 生成 W3C traceparent 字符串。
 * 格式:version-trace_id-parent_id-flags
 */
export function generateTraceparent(): string {
  const version = '00'
  const traceId = randomBytes(16).toString('hex')      // 32 hex
  const parentId = randomBytes(8).toString('hex')       // 16 hex
  const flags = '01'                                    // sampled
  return `${version}-${traceId}-${parentId}-${flags}`
}

/**
 * 从 Fastify request 的 traceparent 头解析 trace 上下文。
 * 如果头不存在或格式非法,返回 null。
 */
export function getTraceparentFromRequest(request: FastifyRequest): TraceContext | null {
  const header = request.headers['traceparent'] as string | undefined
  if (!header) return null
  return parseTraceparent(header)
}

/**
 * 解析 traceparent 字符串。
 * 格式:version-trace_id-parent_id-flags
 */
export function parseTraceparent(traceparent: string): TraceContext | null {
  const parts = traceparent.split('-')
  if (parts.length !== 4) return null
  const [version, traceId, parentId, flags] = parts
  if (!version || !traceId || !parentId || !flags) return null
  if (traceId.length !== 32 || parentId.length !== 16) return null
  if (!/^[0-9a-f]+$/i.test(traceId) || !/^[0-9a-f]+$/i.test(parentId)) return null
  return { version, traceId, parentId, flags }
}

/**
 * 提取 trace_id(32 hex)。
 */
export function extractTraceId(traceparent: string | undefined | null): string | null {
  if (!traceparent) return null
  const ctx = parseTraceparent(traceparent)
  return ctx?.traceId ?? null
}

/**
 * 把 traceparent 注入到出站请求 headers。
 * 用于 api 调用 ai-service 时透传 trace 上下文。
 *
 * 用法:
 *   const headers = propagateToHeaders(request)
 *   const resp = await fetch('http://ai-service:8803/api/llm/complete', { headers })
 *
 * 如果入站 request 无 traceparent,生成新的(开启新 trace)。
 * 如果有,透传原 traceparent(延续同一 trace)。
 */
export function propagateToHeaders(request?: FastifyRequest | null): Record<string, string> {
  const ctx = request ? getTraceparentFromRequest(request) : null
  const traceparent = ctx
    ? `${ctx.version}-${ctx.traceId}-${ctx.parentId}-${ctx.flags}`
    : generateTraceparent()
  return {
    'traceparent': traceparent,
    'X-Trace-Id': ctx?.traceId ?? (traceparent.split('-')[1] ?? ''),
  }
}

/**
 * 生成子 span 的 traceparent(用于内部调用链)。
 * 保持同一 trace_id,生成新的 parent_id。
 */
export function childTraceparent(parent: TraceContext): string {
  const version = parent.version
  const traceId = parent.traceId
  const newParentId = randomBytes(8).toString('hex')
  const flags = parent.flags
  return `${version}-${traceId}-${newParentId}-${flags}`
}

File: apps/ai-service/app/middleware/trace_context.py
"""ai-service traceparent 解析中间件(2026-07-22 立,接收 api 端透传的 trace 上下文)。

与 api 端 utils/trace-context.ts 对等:
- 解析 W3C traceparent 头
- 把 trace_id 注入到 OTel span(关联 api 端 trace)
- 把 trace_id 存入 request.state(供下游使用)
"""

import logging
from typing import Optional
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request

logger = logging.getLogger(__name__)


def parse_traceparent(traceparent: str) -> Optional[dict]:
    """解析 W3C traceparent 字符串。

    格式:version-trace_id-parent_id-flags
    例:00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01

    Returns:
        {"version", "trace_id", "parent_id", "flags"} 或 None
    """
    if not traceparent:
        return None
    parts = traceparent.split("-")
    if len(parts) != 4:
        return None
    version, trace_id, parent_id, flags = parts
    if len(trace_id) != 32 or len(parent_id) != 16:
        return None
    if not all(c in "0123456789abcdef" for c in trace_id.lower()):
        return None
    if not all(c in "0123456789abcdef" for c in parent_id.lower()):
        return None
    return {
        "version": version,
        "trace_id": trace_id,
        "parent_id": parent_id,
        "flags": flags,
    }


class TraceContextMiddleware(BaseHTTPMiddleware):
    """解析 traceparent 头,存入 request.state.trace_id。

    - 如果请求带 traceparent 头:解析并存入 request.state.trace_id
    - 如果不带:不生成新的(api 端负责生成,ai-service 只接收)
    - 不阻塞请求(解析失败也不报错)
    """

    async def dispatch(self, request: Request, call_next):
        traceparent = request.headers.get("traceparent")
        ctx = parse_traceparent(traceparent) if traceparent else None

        if ctx:
            request.state.trace_id = ctx["trace_id"]
            request.state.trace_parent_id = ctx["parent_id"]
            logger.debug("trace_id=%s parent_id=%s", ctx["trace_id"], ctx["parent_id"])
        else:
            request.state.trace_id = None
            request.state.trace_parent_id = None

        response = await call_next(request)

        # 响应头回传 trace_id(便于客户端关联)
        if ctx:
            response.headers["X-Trace-Id"] = ctx["trace_id"]

        return response


def setup_trace_context_middleware(app):
    """注册 trace 上下文中间件到 FastAPI app。"""
    app.add_middleware(TraceContextMiddleware)
